code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -905,77 +912,122 @@ pub mod pallet {
 			(fnft_collection_id, fnft_instance_id): Self::PositionId,
 			amount: Self::Balance,
 			keep_alive: bool,
-		) -> Result<Self::PositionId, DispatchError> {
-			let mut stake = Stakes::<T>::get(fnft_collection_id, fnft_instance_id)
-				.ok_or(Error::<T>::StakeNotFound)?;
-			let mut rewards_pool = RewardPools::<T>::try_get(stake.reward_pool_id)
-				.map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+		) -> DispatchResult {
+			Stakes::<T>::try_mutate(fnft_collection_id, fnft_instance_id, |maybe_stake| {
+				let stake = maybe_stake.as_mut().ok_or(Error::<T>::StakeNotFound)?;
 
-			let reward_multiplier = FixedU64::one().try_into_validated().expect(">= 1");
+				RewardPools::<T>::try_mutate(stake.reward_pool_id, |maybe_rewards_pool| {
+					let rewards_pool =
+						maybe_rewards_pool.as_mut().ok_or(Error::<T>::RewardsPoolNotFound)?;
 
-			ensure!(
-				matches!(
-					T::Assets::can_withdraw(rewards_pool.asset_id, who, amount),
-					WithdrawConsequence::Success
-				),
-				Error::<T>::NotEnoughAssets
-			);
+					ensure!(
+						matches!(
+							T::Assets::can_withdraw(rewards_pool.asset_id, who, amount),
+							WithdrawConsequence::Success
+						),
+						Error::<T>::NotEnoughAssets
+					);
 
-			let awarded_shares = Self::boosted_amount(reward_multiplier, amount)?;
+					let new_shares = Self::boosted_amount(
+						// SAFETY: The duration preset on an existing stake should be valid in the
```

Comment:
I'm afraid we can not make this assumption as it risks throwing off the math for rewarding all users in the pool if we make a mistake in the future. So we are forced to store the `duration_preset` in the `Stake`.

Reply:
Not sure I follow - do we need to store the `reward_multiplier` in the stake as well?

Reply:
Yeah as we now use that for the extension and the risk of any mutation to that is the math being wrong for the entire pool

Reply:
The only way this would error is if the duration presets in the pool changed somehow - if that functionality were added in the future and this is no longer upheld, then the tests would catch it since `defensive*()` methods panic in debug mode when they're triggered. I can replace this with an `expect` call, I just figured a fallback was the best option